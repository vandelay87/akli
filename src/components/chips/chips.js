import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Ripple from '../../hooks/ripple';
import { robotoRegular } from '../../styles/fonts';

const openLink = url => () => window.open(url);
const handleChipKeyPress = event => action => {
  event.preventDefault();
  
  return event.key === 'Enter' && typeof action === 'function' 
    ? action() : openLink(action)()
};

const Chips = ({ list }) => (
  <StyledWrapper>
    {list.map(chip => (
      <StyledChip 
        role="button" 
        tabIndex={0} 
        onClick={chip.link ? openLink(chip.link) : chip.action}
        onKeyPress={(e) => handleChipKeyPress(e)(chip.link || chip.action)}
        key={chip.id}
      >
        <span>{chip.value}</span>
        <Ripple color="gray" duration={600} />
      </StyledChip>
    ))}
  </StyledWrapper>
);

const StyledWrapper = styled.div`
  display: inline-flex;
  flex-wrap: wrap;
  align-items: start;
`;
const StyledChip = styled.div`
  flex: 0 1 auto;
  display: flex;
  align-items: center;
  background: #e0e0e0;
  height: 32px;
  padding: 0 8px;
  border-radius: 16px;
  overflow: hidden;
  position: relative;
  margin: 4px;

  &:hover {
    background: #c2c2c2;
    cursor: pointer;
  }

  &:focus {
    background: #c2c2c2;
    outline: none;
  }

  span {
    ${robotoRegular}
    font-size: .875rem;
    line-height: 1.25rem;
    font-weight: 400;
    letter-spacing: .0178571429em;
  }
`;

Chips.propTypes = {
  list: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    value: PropTypes.string,
    action: PropTypes.func,
    link: PropTypes.string,
  })).isRequired,
};

export default Chips;
