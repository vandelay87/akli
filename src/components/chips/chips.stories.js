import React from 'react';
import { object } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import Chips from './chips';
import { chipsWithLinkMock } from '../../../__mocks__/chips-mock';

export default {
  component: Chips,
  title: 'Chips',
};

export const chipsWithAction = () => {
  const knob = object('URL', [{
    id: '1',
    value: 'Small',
    action: action('small-click'),
  },{
    id: '2',
    value: 'Medium',
    action: action('medium-click'),
  },{
    id: '3',
    value: 'Large',
    action: action('large-click'),
  }]);

  return (
    <Chips list={knob} />
  );
};

export const chipsWithLink = () => (<Chips list={chipsWithLinkMock} />);