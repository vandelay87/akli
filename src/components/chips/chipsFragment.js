import { graphql } from 'gatsby'

export const query = graphql`
  fragment Chips on ContentfulChips {
    id
    list {
      id
      value
      link
      category
    }
  }
`
