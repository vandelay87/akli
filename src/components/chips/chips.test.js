import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Chips from './chips';
import { chipsWithAction, chipsWithLink } from './chips.stories';

describe('chips.js', () => {
  it('renders component', () => {
    const chipsActionTree = renderer
      .create(chipsWithAction())
      .toJSON();
    const chipsLinkTree = renderer
      .create(chipsWithLink())
      .toJSON();

    expect(chipsActionTree).toMatchSnapshot();
    expect(chipsLinkTree).toMatchSnapshot();
  });

  it('calls function', () => {
    const mockFn = jest.fn();
    const wrapper = mount(<Chips list={[{id: '1', value: 'chip', action: mockFn}]} />);

    wrapper.find('div[role="button"]').simulate('click');
    expect(mockFn).toHaveBeenCalledTimes(1);
    wrapper.find('div[role="button"]').simulate('keypress', {key: 'Enter'});
    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  it('opens url', () => {
    window.open = jest.fn();
    const wrapper = mount(chipsWithLink());

    wrapper.find('div[role="button"]').first().simulate('click');
    expect(window.open).toHaveBeenCalledTimes(1);
    wrapper.find('div[role="button"]').first().simulate('keypress', {key: 'Enter'});
    expect(window.open).toHaveBeenCalledTimes(2);
  });
});
