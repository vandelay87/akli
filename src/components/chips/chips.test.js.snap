// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chips.js renders component 1`] = `
<div
  className="sc-bwzfXH gUUAHO"
>
  <div
    className="sc-htpNat dyWzqD"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      Small
    </span>
    <div
      className="sc-bdVaJa esPUWU"
      color="gray"
      onMouseDown={[Function]}
    />
  </div>
  <div
    className="sc-htpNat dyWzqD"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      Medium
    </span>
    <div
      className="sc-bdVaJa esPUWU"
      color="gray"
      onMouseDown={[Function]}
    />
  </div>
  <div
    className="sc-htpNat dyWzqD"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      Large
    </span>
    <div
      className="sc-bdVaJa esPUWU"
      color="gray"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;

exports[`chips.js renders component 2`] = `
<div
  className="sc-bwzfXH gUUAHO"
>
  <div
    className="sc-htpNat dyWzqD"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      React.js
    </span>
    <div
      className="sc-bdVaJa esPUWU"
      color="gray"
      onMouseDown={[Function]}
    />
  </div>
  <div
    className="sc-htpNat dyWzqD"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      Vue
    </span>
    <div
      className="sc-bdVaJa esPUWU"
      color="gray"
      onMouseDown={[Function]}
    />
  </div>
  <div
    className="sc-htpNat dyWzqD"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      Angular
    </span>
    <div
      className="sc-bdVaJa esPUWU"
      color="gray"
      onMouseDown={[Function]}
    />
  </div>
  <div
    className="sc-htpNat dyWzqD"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      Gatsby
    </span>
    <div
      className="sc-bdVaJa esPUWU"
      color="gray"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;
